packages/app-form-builder/src/admin/components/FormEditor/Tabs/EditTab/FormStep/FormStep.tsx

```diff
@@ -81,57 +97,90 @@ export const FormStep = ({
                 // pos.formStep is a source step from which we move row.
                 // formStep is a target step in which we move row.
-                moveRow(pos.row, position.row, formStep.id, sourceStep);
+                moveRow(source.position.row, destination.position.row, source, destination);
                 return;
             }
 
-            // If source pos is set, we are moving an existing field.
-            if (pos) {
-                if (pos.index === null) {
+            if (source.position) {
+                if (source.position.index === null) {
                     console.log("Tried to move Form Field but its position index is null.");
                     console.log(source);
                     return;
                 }
-                // Here we are getting field from the source step ("source step" is a step from which we take a field)
-                const fieldId = sourceStep.layout[pos.row][pos.index];
-                moveField({
-                    field: fieldId,
-                    position,
-                    targetStepId: formStep.id,
-                    sourceStepId: sourceStep.id
-                });
+                const sourceContainer = data.steps.find(step => step.id === source.containerId);
+                const fieldId = sourceContainer?.layout[source.position.row][source.position.index];
+                if (!fieldId) {
+                    console.log("Missing data when moving field.");
+                    return;
+                }
+                moveField({ field: fieldId, target, source, destination });
                 return;
             }
 
             // Find field plugin which handles the dropped field type "name".
-            const plugin = getFieldPlugin({ name });
+            const plugin = getFieldPlugin({ name: target.name });
             if (!plugin) {
                 return;
             }
-            insertField(plugin.field.createField(), position, formStep.id);
+            insertField({
+                data: plugin.field.createField(),
+                target,
+                destination
+            });
         },
         [data]
     );
 
+    const composeHandleDropParams = ({
+        item,
+        destinationPosition
+    }: {
+        item: DragObjectWithFieldInfo;
+        destinationPosition: DropDestination["position"];
+    }) => {
+        handleDrop({
+            target: {
+                type: item.ui,
+                id: item.id,
+                name: item.name
+            },
+            source: {
+                containerId: item?.container?.id,
+                containerType: item?.container?.type,
+                position: item.pos
+            },
+            destination: {
+                containerId: formStep.id,
+                containerType: "step",
+                position: destinationPosition
+            }
+        });
+    };
```

Comment:
Let's have this function extracted into a separate file. Like `handleDrop.ts`?

Oooooooor.... maybe even let's create `useFormStep` hook, which exposes all of these little functions.

Apart from this, I can see other inline stuff below, for example: 

```tsx
onDrop={item => {
    composeHandleDropParams({
        item,
        destinationPosition: {
            row: index,
            index: fieldIndex
        }
    });
    return undefined;
}}
```
Can we have these callbacks returned from the mentioned `useFormStep` hook?

It's just about readability. A lot of inline stuff happening in this file, and I'd say having these extracted in a hook would make the whole file a bit easier to read.